### ISSUE TITLE
`sqrtmod(ZZ(-7), ZZ(1024))` hangs forever (?)

### ISSUE THREAD

**Issue by fingolfin** [Nemocas/Nemo.jl#1254]
```
julia> sqrtmod(ZZ(-7), ZZ(16))
3

julia> sqrtmod(ZZ(-7), ZZ(256))
53

julia> sqrtmod(ZZ(-7), ZZ(512))
75

julia> sqrtmod(ZZ(-7), ZZ(1024))
^C^C^C^C^C^C^C^CWARNING: Force throwing a SIGINT
ERROR: InterruptException:
Stacktrace:
 [1] sqrtmod(x::fmpz, m::fmpz)
   @ Nemo ~/Projekte/OSCAR/Nemo.jl/src/flint/fmpz.jl:918
 [2] top-level scope
   @ REPL[42]:1

julia> 
```
Inspired by https://github.com/oscar-system/Oscar.jl/issues/437

**Comment by wbhart** [comment]
The flint code could/should bomb out after a certain number of iterations and raise an exception. It looks like a flint bug actually.

**Comment by fingolfin** [comment]
Of course in this case, it really should simply find the correct result and return it after an indeed very small number of iterations. I mean, I can do it by hand... :-).
```
julia> S = ResidueRing(ZZ, ZZ(1024))
Integers modulo 1024

julia> S(75 + 256*11)^2 == S(-7)
true
```

But it's even worse:
```
julia> sqrtmod(ZZ(-7), ZZ(2048))
ERROR: no square root exists
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] sqrtmod(x::fmpz, m::fmpz)
   @ Nemo ~/Projekte/OSCAR/Nemo.jl/src/flint/fmpz.jl:920
 [3] top-level scope
   @ REPL[2]:1
```

yet of course `sqrtmod(ZZ(-7), ZZ(2^k))` exists for any `k` ! The rings are so small, I can brute force it:
```
julia> n=2^10 ; S = ResidueRing(ZZ, ZZ(n)) ; findfirst(k->S(k)^2 == S(-7), 1:n)
181

julia> n=2^11 ; S = ResidueRing(ZZ, ZZ(n)) ; findfirst(k->S(k)^2 == S(-7), 1:n)
181

julia> n=2^12 ; S = ResidueRing(ZZ, ZZ(n)) ; findfirst(k->S(k)^2 == S(-7), 1:n)
181
```

**Comment by fingolfin** [comment]
Aha, the docstring for `sqrtmod` says that only prime moduli are supported. So I guess that explain it.

But I don't understand why this limitation exists? Shouldn't crt plus hensel lifting easily take care of the general case?

**Comment by tthsqe12** [comment]
The limitation exists because sqrtmod simply calls the flint function http://flintlib.org/doc/fmpz.html#c.fmpz_sqrtmod and no flint function is going to factor an integer unless you explicitly ask it to.

**Comment by tthsqe12** [comment]
@wbhart The only bug here is that the error message should be "no square root exists or the modulus is not prime"

**Comment by tthsqe12** [comment]
In general, if one is willing to factor the modulus, the function http://flintlib.org/doc/fmpz_mod_poly_factor.html#c.fmpz_mod_poly_roots_factored can take care of square roots.

**Comment by wbhart** [comment]
@tthsqe12 I don't think it gives an error message. It currently just loops forever. Flint could be improved in two ways without factorisation:

1) Observe the bound on the number of iterations in Tonelli-Shanks.
2) Check if the modulus is a perfect power and use lifting.

The only problem is that if the user passes a composite modulus which is not a perfect power, they can get an incorrect answer in finite time.

**Comment by thofma** [comment]
At the moment I don't know how the function handling every modulus should be called. Maybe we should make Nemo handle `sqrtmod` all moduli and have a function `sqrtmod_prime` (there is probably a better name) which assumes the input is a prime? Or maybe a `isprime = true` keyword argument. What y'all think?

**Comment by fingolfin** [comment]
I like having renaming thus function and having a new sqrtmod which factors the modulus. That way naive users can work and experts can still write the most efficient code

**Comment by tthsqe12** [comment]
I actually don't like this sqrtmod_prime approach. It is not hard to create the fields and rings Z/pZ and Z/nZ, and the user should be asking for the normal sqrt in these rings. This rings either know that the modulus is prime by type or can cache the factorization.

**Comment by thofma** [comment]
What are you suggesting? Removing `sqrtmod`?

**Comment by tthsqe12** [comment]
I am suggesting that is not super important how we fix sqrtmod because the sqrt function on gfp_fmpz_elem/fmpz_mod has all of the semantics that one would want.

**Comment by fingolfin** [comment]
Well for the people who run into the gotcha it does matter. IMHO a function named `sqrtmod` should either support all moduli; or explicitly error out for invalid moduli; or not exist at all. I can live with any of these, but the current situation is bad.

**Comment by wbhart** [comment]
It's not possible to support all moduli or error out for invalid moduli without really dramatically changing the run time. (A factor of 5 for a 60 bit prime modulus and a factor of over 2000 for a 200 bit prime modulus.)

I would agree with you if it was called sqrt, but its a low level function. 

What we can do cheaply is error out if it exceeds known bounds on the number of iterations in the algorithm for prime moduli. However, this will not and cannot error out for all invalid moduli. It can return the incorrect answer without that being detected.

I suppose one could check the result is actually a square root before returning it and raise an exception if not. But it will then sometimes work for moduli that it at other times does not work for. Sadly it can't then be used to compute a factorisation of the modulus if it fails, because that is what you are going to need, and that can be very expensive.

**Comment by wbhart** [comment]
And it is documented in Nemo as follows, so it is not as if the behaviour is a surprise:

"Return a square root of x mod m if one exists. The remainder will be in the range [0, m). We require that m is prime, otherwise the algorithm may not terminate."

**Comment by thofma** [comment]
No one said that we don't want the specific functionality of the current `sqrtmod` function with its input restrictions. We don't want to get rid of this.

The problem that I see is that we would also like to have  the general version which can call factoring etc. I think it is expected to be there. But what should it be called?

**Comment by tthsqe12** [comment]
How about we fix the hang inside fmpz_sqrtmod, then Nemo can call this and check its answer. If the answer doesn't checkout, then we factor the modulus and call the appropriate root function, which will give a definitative answer. Good?

**Comment by thofma** [comment]
I like the suggestion from @tthsqe12.

**Comment by wbhart** [comment]
Yeah that seems like an ok solution if it could be made to work. But what should Flint do if the iteration bound is exceeded? It has to do something to prevent the hang and alert the user.

**Comment by tthsqe12** [comment]
Since we are not assuming the input modulus is prime, the fmpz_sqrtmod function's output has to be checked since it is completely unreliable in the non-prime case. Is it possible to change fmpz_sqrtmod so that it still retains the documented behaviour on prime input, and doesn't hang (and gives a 'best guess' for a square root) on non-prime input?

**Comment by wbhart** [comment]
The times that it gives a correct answer for non-prime modulus are exceedingly rare. It can happen, but it's very far from the generic case.

It is possible to retain the existing behaviour in the prime case and have it not hang in the non-prime case (unless I got something wrong somewhere).

The problem is simply what to do (in Flint) when we detect that it was going to hang in the non-prime case. Should we just return random results or raise an exception? If we raise an exception, Nemo would have to be happy with that in the non-prime case. In the fmpz case we could possibly return a flag, but not in the "nmod" case.

**Comment by tthsqe12** [comment]
according to the fmpz_sqrtmod docs, "If 𝑝 is not prime and the return value is 1, the value of 𝑏 is meaningless", so I would return 1 when you don't know what to do (assuming it still works for prime input).

**Comment by wbhart** [comment]
Yeah I just looked at the n_sqrtmod code and we just return 0 there if it would hang (the issue is actually already fixed there). It also just returns 0 if the Jacobi test fails. 

So it looks like we can just fix fmpz_sqrtmod code and not bother setting the output value to anything in particular when it would otherwise hang. But I think we should return 0 in that case for consistency.

Honestly, when the modulus is not prime the return value is pretty useless I'd say. We just have to check the output in Nemo.

Anyway, unless you object, I will go ahead and make this change in Flint and we can make a patch to be merged when Flint-2.9 comes out.
